docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,194 @@
+# ADR 017: Subspace fee grant
+
+## Changelog
+
+- Nov 14th, 2022: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR introduces a new subspace-based fee grant method, which allows subspace service providers to pay the fees of subspace related transactions for the users.
+
+## Context
+
+Currently, one of the major problems of current Web3 services is complicated to use, it requires users to have tokens paying transaction fees so that they can use the service. For instance, in order to create a post on Desmos-based social network, users have to understand:
+- what `transaction` is and why they need some DSM to broadcast it;
+- get some DSM either via an on-ramp or by swapping existing funds.
+
+The `x/feegrant` gives the possibility to pay fees for the users, meaning that users can use the service without understanding how Web3 works behind. However, the `x/feegrant` allowance is not subspace-specified so subspace fees providers might unexpected pay the fees of the transactions inside another subspace.
```

Comment:
I think another thing we should allow is granting allowances to **groups** of users. This way a subspace owner would not need to: 
1. issue an amount of grants equals to the number of active users 
2. remove feegrant allowances as soon as a user becomes inactive/is banned/etc.

Reply:
The idea is so nice, it is so useful for the subspace owner to manage the allowance of each several groups.

Reply:
Using group allowance will have a problem that the fees provider can't set spend and periodic limit for each users since the recorded target is group id. It means that all the users in a group shares the same granted amount if the limit is set.

To solve this problem, Desmos needs to create spend records for all users when a group is granted. Besides, after executing `AddUserToGroup`, `RemoveUserFromGroup`, `DeleteGroup` and `GrantGroupAllowance` actions, desmos are required to update the spend records for all users.

Reply:
> Using group allowance will have a problem that the fees provider can't set spend and periodic limit for each users since the recorded target is group id. It means that all the users in a group shares the same granted amount if the limit is set.

I don't think this is necessarily a problem. It's just how it's supposed to work: if you set a group limit, the limit is shared among the overall group. What owners can do is setting a time limit, but not an amount limit. Or, simply, they can set a very high amount limit instead. That being said, I think it's still something we should probide since it would simplify the grant management a lot. 

> In addition, a user might have several allowances by joining different groups, it also increases the complexity.

I don't think it's extremely complicated for the following reasons: 
1. user groups will not be that many inside a subspace.  
   I don't see a reason why a subspace would have hundreds or thousands of user groups: it just makes it really hard for the owner to manage them properly. 
2. even if the subspace has millions of groups, it doesn't matter. We will iterate *first* on the grants given to various groups, and then check if either one of those groups contains the user. So we scale based on the number of grants, instead of on the number of groups. And, since groups are hierarchical (like permissions), we can stop as soon as we find the first group

> Any thoughts?

Reply:
Got you! The complicated thingy only happens if desmos records all users in the granted group but it is not necessarily. The ADR group allowance will be all the users share the group limit.